tensorflow/core/graph/graph_constructor_test.cc

```diff
@@ -3205,6 +3205,49 @@ TEST_F(GraphConstructorTest, ImportGraphDef_ValidateColationConstraints) {
   TF_EXPECT_OK(ImportGraphDef(options, def, &graph_, nullptr));
 }
 
+TEST_F(GraphConstructorTest, ImportGraphDef_ValidateDefaultDevice) {
+  std::string stored_dev = "some dev";
+  std::string gdef_ascii(
+      R"EOF(
+      node { name: 'test_input' op: 'TestInput' }
+      node { name: 'test_input_with_dev' op: 'TestInput' device: 'some dev'}
+      node { name: 'test_op' op: 'TestMul' input: [ 'test_input:0', 'test_input:1' ] }
+      node { name: 'test_op_with_dev' op: 'TestMul' input: [ 'test_input:0', 'test_input:1' ] device: 'some dev'}
+      )EOF");
+
+  GraphDef gdef;
+  CHECK(protobuf::TextFormat::ParseFromString(gdef_ascii, &gdef));
+
+  ImportGraphDefOptions options;
+  // assign this execution device for the nodes, which do not alreadt have default device, specified in graph_def
+  // node->requested_device() should be either this assigned device, or default device from graph_def
+  std::string assigned_dev = "/gpu:13";
+  options.default_device = assigned_dev;
+
+  std::map<std::string, std::string> node_dev_map;
+  std::vector<std::string> assigned_dev_nodes = {"test_input", "test_op"};
+  std::vector<std::string> stored_dev_nodes = {"test_input_with_dev", "test_op_with_dev"};
+  options.return_nodes = assigned_dev_nodes;
+  options.return_nodes.insert(options.return_nodes.end(), stored_dev_nodes.begin(), stored_dev_nodes.end() );
+
+  for (auto node: assigned_dev_nodes) {
```

Comment:
The test will be easier to follow if we explicitly compare devices instead of having to sort through this `node_dev_map` (and whether names are overridden etc.). For example, we could  do something like this:

```c++
GraphDef gdef;
ASSERT_TRUE(protobuf::TestFormat::ParseFromString(gdef_ascii, &gdef));

ImportGraphDefOptions options;
options.default_device = "/gpu:13";
ImportGraphDefResults res;

TF_ASSERT_OK(ImportGraphDef(options, &def, &graph_, NULL, &res));
std::map<string, string> node2dev;
for (Node* n : graph_.nodes()) {
  node2dev[n->name()] = n->assign_device_name();
}
EXPECT_EQ(node2dev["test_input"], "/gpu:13");
EXPECT_EQ(node2dev["test_op"], "/gpu:13");
EXPECT_EQ(node2dev["test_input_with_dev", "some dev");
EXPECT_EQ(node2dev["test_op_with_dev", "some dev");
```

Reply:
This particular test tests different device, node can only be assigned from graph_def, if every node there has a device specification, which is not the case generally. This will work if we check `requested_device` though, I've updated the test, please check